### PR TITLE
RBAC: Add scope resolvers for dashboards

### PR DESCRIPTION
**Pull request by kalleep**
**What this PR does / why we need it**:
Inject access control into dashboard service and register scope resolvers for both id and uid

The resolvers will resolve a dashboard scope into uid based scope for dasboard and the folder the dashboard belongs to.

With this change we can:
* Remove adition folder checks in access control guardian
* Remove the quick fix in dashboard permission service

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/3397

**Special notes for your reviewer**:
* Most of the changes is in tests where the dashboard service is used

### CODE REVIEW

File: pkg/services/accesscontrol/scope.go [hunk 1]
```diff
@@ -2,13 +2,22 @@ package accesscontrol
...
 	maxPrefixParts = 2
 )
 
+func ParseScopeID(scope string) (int64, error) {
+	return strconv.ParseInt(scope[len(ScopePrefix(scope)):], 10, 64)
+}
+
+func ParseScopeUID(scope string) (string, error) {
+	return scope[len(ScopePrefix(scope)):], nil
+}
```
Comment (gamab) [inline pkg/services/accesscontrol/scope.go:19]:
```suggestion
func ParseScopeID(scope string) (int64, error) {
	return strconv.ParseInt(ScopeSuffix(scope), 10, 64)
}

func ScopeSuffix(scope string) string {
	return scope[len(ScopePrefix(scope)):]
}
```

File: pkg/services/dashboards/accesscontrol.go [hunk 2]
```diff
@@ -81,3 +82,66 @@ func NewFolderIDScopeResolver(db Store) (string, ac.ScopeAttributeResolver) {
...
+		dashboard, err := db.GetDashboard(ctx, &models.GetDashboardQuery{Uid: uid, OrgId: orgID})
+		if err != nil {
+			return nil, err
+		}
+
+		return resolveDashboardScope(ctx, db, orgID, dashboard)
+	})
+}
+
+func resolveDashboardScope(ctx context.Context, db Store, orgID int64, dashboard *models.Dashboard) ([]string, error) {
```
Comment (gamab) [inline pkg/services/dashboards/accesscontrol.go:132]: Nit.
```suggestion
func dashboardAndParentScopes(ctx context.Context, db Store, orgID int64, dashboard *models.Dashboard) ([]string, error) {
```

File: pkg/services/dashboards/accesscontrol.go [hunk 3]
```diff
@@ -2,9 +2,10 @@ package dashboards
 
 import (
 	"context"
-	"strconv"
 	"strings"
 
+	"github.com/grafana/grafana/pkg/models"
+
```
Comment (gamab) [inline pkg/services/dashboards/accesscontrol.go:8]:
```suggestion
```

**Review by gamab — APPROVED** [review]

File: pkg/services/dashboards/accesscontrol.go [hunk 4]
```diff
@@ -81,3 +82,66 @@ func NewFolderIDScopeResolver(db Store) (string, ac.ScopeAttributeResolver) {
...
+	prefix := ScopeDashboardsProvider.GetResourceScopeUID("")
+	return prefix, ac.ScopeAttributeResolverFunc(func(ctx context.Context, orgID int64, scope string) ([]string, error) {
+		if !strings.HasPrefix(scope, prefix) {
+			return nil, ac.ErrInvalidScope
+		}
+
+		uid, err := ac.ParseScopeUID(scope)
+		if err != nil {
+			return nil, ac.ErrInvalidScope
+		}
```
[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/dashboards/accesscontrol.go:120]: Yes but we can do the check in ParseScopeUID and return the invalid scope err from there

**Pull request merged**